osu.Game.Rulesets.Catch/Edit/Blueprints/Components/SelectionEditablePath.cs

```diff
@@ -19,22 +20,28 @@ public partial class SelectionEditablePath : EditablePath, IHasContextMenu
     {
         public MenuItem[] ContextMenuItems => getContextMenuItems().ToArray();
 
+        private readonly JuiceStream juiceStream;
+
         // To handle when the editor is scrolled while dragging.
         private Vector2 dragStartPosition;
 
         [Resolved]
         private IEditorChangeHandler? changeHandler { get; set; }
 
-        public SelectionEditablePath(Func<float, double> positionToTime)
+        public SelectionEditablePath(JuiceStream juiceStream, Func<float, double> positionToTime)
             : base(positionToTime)
         {
+            this.juiceStream = juiceStream;
         }
 
         public void AddVertex(Vector2 relativePosition)
         {
+            changeHandler?.BeginChange();
             double time = Math.Max(0, PositionToTime(relativePosition.Y));
             int index = AddVertex(time, relativePosition.X);
+            UpdateHitObjectFromPath(juiceStream);
```

Comment:
Calls are required to actually make undo/redo work, because without this call, nothing in the actual beatmap changes, and `changeHandler?.EndChange()` ends up being a no-op.